src/nbl/video/CVKLogicalDevice.h

```diff
@@ -124,51 +327,472 @@ class CVKLogicalDevice final : public ILogicalDevice
+
+            vk_copyDescriptorSets[i].srcBinding = pDescriptorCopies[i].srcBinding;
+            vk_copyDescriptorSets[i].srcArrayElement = pDescriptorCopies[i].srcArrayElement;
+
+            // if (pDescriptorCopies[i].dstSet->getAPIType() != EAT_VULKAN)
+            //     continue;
+            vk_copyDescriptorSets[i].dstSet = static_cast<const CVulkanDescriptorSet*>(pDescriptorCopies[i].dstSet)->getInternalObject();
+
+            vk_copyDescriptorSets[i].dstBinding = pDescriptorCopies[i].dstBinding;
+            vk_copyDescriptorSets[i].dstArrayElement = pDescriptorCopies[i].dstArrayElement;
+            vk_copyDescriptorSets[i].descriptorCount = pDescriptorCopies[i].count;
+        }
+
+        vkUpdateDescriptorSets(m_vkdev, descriptorWriteCount, vk_writeDescriptorSets, descriptorCopyCount, vk_copyDescriptorSets);
+    }
+
+    core::smart_refctd_ptr<IDriverMemoryAllocation> allocateDeviceLocalMemory(
+        const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs) override
+    {
+        // Todo(achal): Remove this, and integrate it with
+        // IDriverMemoryBacked::SDriverMemoryRequirements if possible. Probably we can cache this
+        // in a struct of 5 uint32_t's (deviceLocal, spillover, upstreaming,
+        // downstreaming, cpuSizeGPUVisible) per CVulkanPhysicalDevice and just access
+        // the relevant index in here.
+#if 0
+        uint32_t memoryTypeIndex = ~0u;
+        {
+            if (m_physicalDevice->getAPIType() != EAT_VULKAN)
+                return nullptr;
+
+            VkPhysicalDevice vk_physicalDevice = static_cast<const CVulkanPhysicalDevice*>(m_physicalDevice.get())->getInternalObject();
+
+            VkPhysicalDeviceMemoryProperties vk_physicalDeviceMemoryProperties;
+            vkGetPhysicalDeviceMemoryProperties(vk_physicalDevice, &vk_physicalDeviceMemoryProperties);
+
+            // Todo(achal): I need to take into account getDeviceLocalGPUMemoryReqs, probably
+            const uint32_t supportedMemoryTypes = additionalReqs.vulkanReqs.memoryTypeBits;
+            IDriverMemoryBacked::SDriverMemoryRequirements memoryReqs = getDeviceLocalGPUMemoryReqs();
+
+            const VkMemoryPropertyFlags desiredMemoryProperties =
+                VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+
+            for (uint32_t i = 0u; i < vk_physicalDeviceMemoryProperties.memoryTypeCount; ++i)
+            {
+                const bool isMemoryTypeSupportedForResource = (supportedMemoryTypes & (1 << i));
+
+                const bool doesMemoryHaveDesirableProperites =
+                    (vk_physicalDeviceMemoryProperties.memoryTypes[i].propertyFlags
+                        & desiredMemoryProperties) == desiredMemoryProperties;
+                if (isMemoryTypeSupportedForResource && doesMemoryHaveDesirableProperites)
+                {
+                    memoryTypeIndex = i;
+                    break;
+                }
+            }
```

Comment:
General outline:
1. use  `getDeviceLocalMemReqs` to get an initial `IDriverMemoryBacked::SDriverMemoryRequirements`
2. patch it up with `additionalReqs` as per my earlier comment (these will usually come from `vkGetBufferMemoryRequirements`)
2. Translate the resulting ``IDriverMemoryBacked::SDriverMemoryRequirements` into `VkMemoryPropertyFlagBits `
3. go through the memory types and check:
- bitfield matches in `supportedMemoryTypes`
- memory properties are fullfilled
- there's enough memory left on the heap to allocate (you can try to allocate on each matching heap, and only quit the loop if allocation was a success)

Reply:
> you can try to allocate on each matching heap, and only quit the loop if allocation was a success

I don't think that is how it works.
After picking a suitable `VkMemoryType` from `VkPhysicalDeviceMemoryProperties::memoryTypes`, say `memoryTypes[i]`, you have to pass the corresponding index (`i`) to `VkMemoryAllocateInfo`, which will try to allocate memory from the heap at index `memoryTypes[i].heapIndex`. Sure, we can check if the heap in question has enough space left preemptively (for example, by enabling `VkPhysicalDeviceMemoryBudgetPropertiesEXT`), but I don't see any point, as:
>  Some platforms may have a limit on the maximum size of a single allocation. For example, certain systems may fail to create allocations with a size greater than or equal to 4GB. Such a limit is implementation-dependent, and if such a failure occurs then the error VK_ERROR_OUT_OF_DEVICE_MEMORY must be returned. This limit is advertised in VkPhysicalDeviceMaintenance3Properties::maxMemoryAllocationSize.

Reply:
For reference:
> VkMemoryRequirements::memoryTypeBits is a bitmask and contains one bit set for every supported memory type for the resource. Bit i is set if and only if the memory type i in the VkPhysicalDeviceMemoryProperties structure for the physical device is supported for the resource.

Reply:
My bad, I think it is possible.
You cannot target a specific heap by index, but you can always find a list of compatible memory types and try with their associated heaps.

Reply:
> You cannot target a specific heap by index, but you can always find a list of compatible memory types and try with their associated heaps.

That's the idea.

Reply:
All the "hardcoded" reqs you get from here are "conservative" (overly lax)
https://github.com/Devsh-Graphics-Programming/Nabla/blob/scene_manager/include/nbl/video/ILogicalDevice.h#L142

So that's why you'd need to "tighten" them up with additionalMemReqs (usually from `vkGetBufferMemoryRequirements`/`vkGetImageMemoryRequirements`), with stuff like:
- constrain the alignment (take the highest value)
- constrain the size (overwrite)
- constrain memory type bits (bitwise-AND them with actual types you can deduce from devicelocal or not and mapping caps like coherent cached etc. as well as the buffer memory requirements)
- overwrite dedicated allocation requirement/preference as needed

Reply:
resolved